miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/MiniAppManifestCache.kt

```diff
@@ -0,0 +1,101 @@
+internal class MiniAppManifestCache(
+    context: Context,
+    val miniAppCustomPermissionCache: MiniAppCustomPermissionCache
+) {
+    private val prefs: SharedPreferences = context.getSharedPreferences(
+        "com.rakuten.tech.mobile.miniapp.manifest.cache", Context.MODE_PRIVATE
+    )
+
+    /**
+     * Reads the grant results from SharedPreferences.
+     * @param [miniAppId] the key provided to find the stored manifest per MiniApp.
+     * @return [MiniAppManifest] an object to contain the manifest per MiniApp.
+     * if data has been stored in cache, otherwise empty value.
+     */
+    fun readMiniAppManifest(miniAppId: String): MiniAppManifest? {
+        val manifestJsonStr = prefs.getString(miniAppId, null) ?: return null
+        return try {
+            Gson().fromJson(manifestJsonStr, object : TypeToken<MiniAppManifest>() {}.type)
+        } catch (e: Exception) {
+            Log.e(this::class.java.canonicalName, e.message.toString())
+            null
+        }
+    }
+
+    /**
+     * Stores [MiniAppManifest] to SharedPreferences.
+     * @param [miniAppId] the key provided to find the stored manifest per MiniApp.
+     * @param [miniAppManifest] an object to contain the manifest values per MiniApp.
+     */
+    fun storeMiniAppManifest(
+        miniAppId: String,
+        miniAppManifest: MiniAppManifest
+    ) {
+        val jsonToStore: String = Gson().toJson(miniAppManifest)
+        prefs.edit().putString(miniAppId, jsonToStore).apply()
+    }
+
+    /**
+     * Returns the list of all manifest permissions e.g. required and optional.
+     * @param [miniAppId] the key provided to find the stored manifest per MiniApp.
+     */
+    fun getCachedAllPermissions(miniAppId: String) = getCachedRequiredPermissions(miniAppId) +
+            getCachedOptionalPermissions(miniAppId)
+
+    /**
+     * Returns true if the required permissions are denied, otherwise false.
+     * @param [miniAppId] the key provided to find the stored manifest per MiniApp.
+     * @param [manifest] the manifest includes the required permissions.
+     */
+    fun isRequiredPermissionDenied(miniAppId: String, manifest: MiniAppManifest?): Boolean {
+        // store manifest value in cache before checking the required permission status
+        if (manifest != null) {
+            storeMiniAppManifest(miniAppId, manifest)
+
+            getCachedRequiredPermissions(miniAppId).find {
+                it.second != MiniAppCustomPermissionResult.ALLOWED
+            }?.let { return true }
+        }
+        return false
+    }
+
+    @VisibleForTesting
+    fun getCachedRequiredPermissions(
+        miniAppId: String
+    ): List<Pair<MiniAppCustomPermissionType, MiniAppCustomPermissionResult>> {
+        val manifest = readMiniAppManifest(miniAppId)
+        val cachedPermissions = miniAppCustomPermissionCache.readPermissions(miniAppId).pairValues
+        return manifest?.requiredPermissions?.mapNotNull { (first) ->
```

Comment:
if `manifest` is null then does `getCachedRequiredPermissions` get crash?

Reply:
Yes right, so I have updated `getCachedRequiredPermissions` & `getCachedOptionalPermissions` to handle this null.